src/libopensc/card-edo.c

```diff
@@ -0,0 +1,299 @@
+		else if (pathlen == 2)
+			LOG_TEST_RET(card->ctx, edo_select_ef(card, path, buff), "EF select failed");
+		else
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+		path += 2;
+		pathlen -= 2;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+/*! Selects file specified by given path.
+ *
+ * Card does not support selecting file at once, that's why it have to be done in following way:
+ * 1. Select AID if provided,
+ * 2. Select MF if provided,
+ * 3. Select DF until provided,
+ * 4. Select EF if provided.
+ */
+static int edo_select_file(struct sc_card* card, const struct sc_path* in_path, struct sc_file** file_out) {
+	LOG_FUNC_CALLED(card->ctx);
+	struct edo_buff buff;
+
+	switch (in_path->type) {
+		case SC_PATH_TYPE_PATH:
+		case SC_PATH_TYPE_FILE_ID:
+			if (in_path->aid.len)
+				LOG_TEST_RET(card->ctx, edo_select_name(card, in_path->aid.value, in_path->aid.len, &buff), "Select AID failed");
+			if (in_path->len)
+				LOG_TEST_RET(card->ctx, edo_select_path(card, in_path->value, in_path->len, &buff), "Select path failed");
+			break;
+		case SC_PATH_TYPE_DF_NAME:
+			LOG_TEST_RET(card->ctx, edo_select_name(card, in_path->value, in_path->len, &buff), "Select AID failed");
+			break;
+		default:
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	}
+
+	if (file_out) {
+		if (buff.len < 2)
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_UNKNOWN_DATA_RECEIVED);
+		if (!(*file_out = sc_file_new()))
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+		(*file_out)->path = *in_path;
+		LOG_TEST_RET(card->ctx, card->ops->process_fci(card, *file_out, buff.val, buff.len), "Process FCI failed");
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+/*! Computes ECDSA signature.
+ *
+ * If ECDSA was used, the ASN.1 sequence of integers R,S returned by the
+ * card needs to be converted to the raw concatenation of R,S for PKCS#11.
+ */
+static int edo_compute_signature(struct sc_card* card, const u8* data, size_t datalen, u8* out, size_t outlen) {
+	LOG_FUNC_CALLED(card->ctx);
+	struct edo_buff buff = {{}, sizeof buff.val};
+
+	LOG_TEST_RET(card->ctx, sc_get_iso7816_driver()->ops->compute_signature(card, data, datalen, buff.val, buff.len), "Internal signature failed");
+
+	LOG_TEST_RET(card->ctx, sc_asn1_sig_value_sequence_to_rs(card->ctx, buff.val, buff.len, out, outlen), "ASN.1 conversion failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+/*! Sets security environment
+ *
+ * Card expects key file to be selected first, followed by the
+ * set security env packet with: 0x80, 0x1, 0xcc, 0x84, 0x1, 0x80|x,
+ * where x is the key reference byte.
+ */
+static int edo_set_security_env(struct sc_card* card, const struct sc_security_env* env, int se_num) {
+	LOG_FUNC_CALLED(card->ctx);
+	struct sc_apdu apdu;
+
+	if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN)
+		sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x41, 0xB6, (u8[]) {
+			0x80, 0x1, 0xcc, 0x84, 0x1, 0x80 | env->key_ref[0]
```

Comment:
I think we don't use dynamic data this way in the rest of the code. Please create a seperate data buffer with something like this:
```
u8[] data ={0x80, 0x1, 0xcc, 0x84, 0x1, 0x80};
if (env->flags & SC_SEC_ENV_KEY_REF_PRESENT)
    data[5] |=  env->key_ref[0];
```